tests/analysis/impl/dense/gpcca_test.py

```diff
@@ -0,0 +1,638 @@
+
+        for i, mu_ in enumerate([0, 10, 50, 100, 200, 500, 1000]):
+            mu_ = mu(mu_)
+            P, sd = get_known_input(mu_)
+            X, _, _ = _do_schur(P, eta=sd, m=kmax)
+
+            crisp = [-np.inf] * (kmax - kmin)
+            for j, k in enumerate(range(kmin, kmax)):
+                svecs = X[:, :k]
+                A = _initialize_rot_matrix(svecs)
+
+                _, _, fopt = _opt_soft(svecs, A)
+                crisp[j] = (k - fopt) / k
+                kopt[i] = ks[np.argmax(crisp)]
+
+        np.testing.assert_array_equal(kopt, [3, 3, 3, 3, 2, 2, 7])
+
+    def test_cluster_by_first_col_not_1(self):
+        svecs = np.zeros((4, 3))
+        svecs[0, 0] = 1
+
+        with pytest.raises(
+            ValueError,
+            match="First Schur vector is not constant. This indicates that the Schur vectors are incorrectly sorted. "
+            "Cannot search for a simplex structure in the data",
+        ):
+            _cluster_by_isa(svecs)
+
+    def test_cluster_by_isa_shape_error(self):
+        svecs = np.zeros((3, 4))
+        svecs[:, 1] = 1.0
+
+        with pytest.raises(
+            ValueError,
+            match=r"The Schur vector matrix of shape \(\d+, \d+\) has more columns than rows. You can't get a "
+            r"\d+-dimensional simplex from \d+ data vectors.",
+        ):
+            _cluster_by_isa(svecs)
+
+    def test_cluster_by_isa(
+        self, chi_isa_mu0_n3: np.ndarray, chi_isa_mu100_n3: np.ndarray
+    ):
+        # chi_sa_mu0_n3 has permuted 2nd and 3d columns when compared to the matlab version
+        for m, chi_exp in zip([0, 100], [chi_isa_mu0_n3, chi_isa_mu100_n3]):
+            mu_ = mu(m)
+            P, sd = get_known_input(mu_)
+            X, _, _ = _do_schur(P, sd, m=3)
+            chi, _ = _cluster_by_isa(X[:, :3])
+
+            assert_allclose(chi.T @ chi, chi_exp.T @ chi_exp)
+            assert_allclose(chi, chi_exp)
+
+    def test_use_minChi(self):
+        kmin, kmax = 2, 9
+        kopt = [None] * 7
+
+        for i, m in enumerate([0, 10, 50, 100, 200, 500, 1000]):
+            mu_ = mu(m)
+            P, sd = get_known_input(mu_)
+            g = GPCCA(P, eta=sd)
+            minChi = g.minChi(kmin, kmax)
+
+            kopt[i] = kmax - 1 - np.argmax(np.flipud(minChi[1:-1]))
+
+        np.testing.assert_array_equal(kopt, [3] * 6 + [7])
+
+    def test_gpcca_brandts_sparse_is_densified(self, P: np.ndarray, sd: np.ndarray):
+
+        with pytest.warns(
+            UserWarning,
+            match=r"Sparse implementation is only avaiable for `method='krylov'`, densifying.",
+        ):
+            GPCCA(csr_matrix(P), eta=sd, method="brandts").optimize(3)
+
+
+@skip_if_no_petsc_slepc
+class TestPETScSLEPc:
+    def test_do_schur_krylov(self, example_matrix_mu: np.ndarray):
+        N = 9
+        P, sd = get_known_input(example_matrix_mu)
+
+        X_k, RR_k, _ = _do_schur(P, eta=sd, m=N, method="krylov")
+
+        _assert_schur(P, X_k, RR_k, N)
+
```

Comment:
Is there a reason why we do not test Brandts method with the examples of Brandts and from Bai and Demmel?
See test_SRSchur_num_t at https://github.com/msmdev/gpcca/blob/66b3cb079847328b01fe61361c952f8b1093f3a9/unitTests.m#L3085

Reply:
I must've forgotten, will add it soon.

Reply:
Done here: https://github.com/msmdev/msmtools/pull/18/files#diff-a77e704d8e9642584dac7cea5a986e5cf755403a1056410c6718fc75b012d381R536
@msmdev 3 out of 5 tests fail in this test: R_2, R_3, R_5 (on the eigvalues assertion)
I've double checked whether I've transcribed the matrices, seems correct (please check this also, the matrices are here: https://github.com/msmdev/msmtools/pull/18/files#diff-e52e4ddd58b7ef887ab03c04116e676f6280b824ab7469d5d3080e5cba4f2128R1329)

Reply:
Can't find any of your changes under the supplied links... there seems to be a general problem with all links of this type (also in all above and below cases). Found it searching the commits...
Could you please tell me in more detail what exactly fails?

Reply:
See my https://github.com/msmdev/msmtools/pull/18#issuecomment-736507184